File: p031.py
def count_change_ways(amount, coins):
    return (0 if amount < 0 or not coins else
            1 if amount == 0 else
            (count_change_ways(amount, coins[1:])
             + count_change_ways(amount - coins[0], coins)))

ans = count_change_ways(200, [200, 100, 50, 20, 10, 5, 2, 1])

print(ans)

File: p031.scala
def countChangeWays(amount: Int, coins: Array[Int]): Int =
  if (amount < 0 || coins.length == 0) 0
  else if (amount == 0) 1
  else (countChangeWays(amount, coins.drop(1)) +
    countChangeWays(amount - coins(0), coins))

val ans = countChangeWays(200, Array(200, 100, 50, 20, 10, 5, 2, 1))

println(ans)
